### ISSUE TITLE
Synchronous message handling

### ISSUE THREAD

**Issue by tulir** [Rhymen/go-whatsapp#160]
All the handlers are called in a goroutine, which is problematic when you want to keep message ordering. It's especially a problem for the bunch of messages that WhatsApp sends when connecting.

There should be some option for synchronous handling, maybe calling the handlers without a goroutine (and let the handlers themselves make sure handling doesn't take too long) or a channel where all new events are sent.

**Comment by tulir** [comment]
I'll probably implement some solution for this and make a pull request. @SchulteMK @Rhymen do you have a preference on how this should be implemented?

**Comment by houstondapaz** [comment]
Why not you only sort messages by timeStamp?

**Comment by highness28** [comment]
Create a queue architecture in your back-end. In that way all new messages that passes through your handlers are automatically queued in order. There is a down-side though since the newly updated repository has a bug in handling socket connection to whatsapp, it is not fail safe so if ever you got disconnected to whatsapp socket your go-whatsapp service will not receive any messages.

**Comment by tulir** [comment]
I did create a queue (go channel) architecture, but upstream go-whatsapp calls all handlers in their own goroutines, which means execution order (thus the queue insertion order) is not guaranteed.

**Comment by SchulteMK** [comment]
@tulir you are right, execution order is not guaranteed with multiple goroutines "running" at the same time. Do you have an idea for implementation? I already wanted to change the whole handler design and make it synchronous, but I was not truly happy with my implementation

**Comment by SchulteMK** [comment]
Just came to my mind, that we could dispatch every message in its raw decrypted form and offer helper functions to handle the individual types of messages. But I am not sure if it is a good ideal. 
What do you think? Better have one handler and use helper functions or have handlers for every message type as we have now?

**Comment by SchulteMK** [comment]
I just poked around some forks of go-whatsapp and found [this implementation](https://github.com/dimaskiddo/go-whatsapp/commit/019432405d461ffca30e5888b88fd6fc8052cc8f) by @dimaskiddo.
This could be a simple solution that does not break anything

**Comment by tulir** [comment]
That looks like it uses my implementation (same style and function name, https://github.com/tulir/go-whatsapp/tree/sync-handlers), but I didn't see my commits when looking through quickly

**Comment by dimaskiddo** [comment]
I'am sorry i just made mistake and removed your authored @tulir, i really sorry i just realized it, doesn't mean to do that. And yes it was @tulir code there

Once more sorry @tulir for what i've done

**Comment by codenoid** [comment]
`which is problematic when you want to keep message ordering` i use epoch timestamp from message
